[TOPIC-GPIO] Update gpio_intel_apl driver to new GPIO API

dcpleung opened the pull request:
This updates the `gpio_intel_apl` to the new GPIO API. This also changes how the GPIO controllers are defined in the DTS, to a) remove the entries in `dts_fixup.h`, and b) allow referencing these controllers from other DTS files. Also adds a few board overlay files to the new GPIO API tests and the sample GPIO app.

This has been tested on the `up_squared` board. However, these is a FPGA acting as I/O buffer on the HAT where settings are controlled by BIOS. Pins can be set as input or output as GPIOs. Since the FPGA is actually driving the pins, pull-up/pull-down tests on the input pin do not work correctly.

dcpleung committed:
dts: x86/apollo_lake: update UART and I2C unit address

There were complains about the unit-address and first reg
not matching. So update the DTS file to match.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
gpio: intel_apl: convert to new GPIO API

Update driver code to use new GPIO configuration flags such as
GPIO_ACTIVE_LOW. Also add implementation of new port_* driver
API as well as gpio_pin_interrupt_configure function.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
samples: gpio: add board overlay for up_squared

This adds a board overlay for up_squared to be used with
the GPIO sample. Using the overlay is because the pins
being used are not actually LED and switch but GPIO pins
used as such.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
tests: gpio_basic_api: add board overlay for up_squared

This adds a board overlay for up_squared to be used with
the GPIO basic API test.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
tests: gpio_api_1pin: add board overlay for up_squared

This adds a board overlay for up_squared to be used with
the GPIO basic API test.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

mnkp commented on drivers/gpio/gpio_intel_apl.c:
```diff
@@ -304,6 +284,87 @@ static int gpio_intel_apl_config(struct device *dev, int access_op,
+	raw_pin = cfg->pin_offset + pin;
+
+	if (!check_perm(dev, raw_pin)) {
+		return -EPERM;
```
This error code is not documented in GPIO API. If we need it we should update the API. I'm not familiar with this architecture. What exactly is happening here? Could we use `EINVAL`?

dcpleung replied:
Some of the pins are controlled by, for example, the power management controller. These pins cannot be manipulated. I can use `EINVAL`. Originally `EPERM` is used because the datasheet talks about permissions.

mnkp replied:
I see you've modified the code to use `EINVAL` but the `EPERM` use case also makes sense. The pin is valid, but the driver cannot use it since its permanently assigned to a different peripheral without a possibility to 'steal' it. 

If you prefer `EPERM` feel free to revert the change and I will prepare the PR to update GPIO API. I'll leave it to you to decide.

dcpleung replied:
Let's leave it as `EINVAL` for now. It makes error checking on applications easier, as the pin cannot be configured anyway.

mnkp commented on drivers/gpio/gpio_intel_apl.c:
```diff
@@ -304,6 +284,87 @@ static int gpio_intel_apl_config(struct device *dev, int access_op,
+
+	/* clear (by setting) interrupt status bit */
+	reg = cfg->reg_base + REG_GPI_INT_STS_BASE;
+	sys_bitfield_set_bit(reg, raw_pin);
```
Shouldn't we clear interrupt status after new configuration is applied and shortly before enabling it? Otherwise pin will be sensitive for a while to some old interrupt configuration we have no control of.

mnkp commented on drivers/gpio/gpio_intel_apl.c:
```diff
@@ -304,6 +284,87 @@ static int gpio_intel_apl_config(struct device *dev, int access_op,
+
+	/* write back pad configuration register after all changes */
+	sys_write32(cfg0, reg);
+	sys_write32(cfg1, (reg + 4));
```
We could skip extra parenthesis, doesn't improve readability:
```
	sys_write32(cfg1, reg + 4);
```

mnkp reviewed (commented):
LGTM with minor comments.

mnkp reviewed (approved):
I reviewed the driver code. I didn't review the DTS since I'm not familiar with the architecture.

carlescufi commented:
@galak @pabigot can we merge this?

pabigot reviewed (commented):
I don't have any objections merging this for the topic branch, though the statement that pulls can't be verified (though they're supported) is a bit worrisome.

mnkp commented:
It would be good if @galak had a quick look and verified DTS usage. It's a bit non standard.

galak reviewed (changes requested):
Yeah, the DTS looks a little odd,  Hopefully @dcpleung can point me at a manual/specs for the SoC to understand the register layout and why we need 'offset' and multiple GPIO controller nodes.

dcpleung commented:
> Yeah, the DTS looks a little odd, Hopefully @dcpleung can point me at a manual/specs for the SoC to understand the register layout and why we need 'offset' and multiple GPIO controller nodes.

There are 4 GPIO controllers on the SoC, where each one has more than 32 pins. Since the GPIO API is based on the assumption that one port has 32 pins, this driver has to divide one controller into a few driver instances (hence the pin offsets).

The datasheet is @ https://www.intel.com/content/www/us/en/processors/pentium/pentium-celeron-n-series-j-series-datasheet-vol-3.html, and GPIO is chapter 22. The section 22.[1-4] describe the register sets for the 4 GPIO controllers.

pabigot committed:
x86: apollo_lake: add nexus map for GPIO nodes

This maps devicetree GPIO phandle arrays from the full controller pin
range to the sub-controllers required by Zephyr's limit of 32 pins per
controller device.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot commented:
As suggested in the API telecon: https://github.com/pabigot/zephyr/commits/pr/20400 provides a nexus map that can be used to translate GPIO references from the controller to subdevices.  I.e. devicetree source can use `&gpio_w 42 0` instead of `&gpio_w_032_046 10 0`.

If you find this useful you should cherrypick the map and squash the overlay updates.  The bash script used to generate the bulk of the map may be kept for reference.

dcpleung commented:
@pabigot Thank you. I have updated the PR with your changes. :)

galak commented on dts/bindings/gpio/intel,apl-gpio.yaml:
```diff
@@ -20,6 +20,14 @@ properties:
+    ngpios:
+      required: true
+
+    offset:
```
Can we rename this `pin-offset`

dcpleung replied:
Changed, and updated.

dcpleung committed:
x86: apollo_lake: changes GPIO definition in DTS

This breaks down the GPIO controller definition in DTS into
multiple entries. This allows these controllers to be
referenced by other DTS, and test board overlay files.
And also we can remove the entries in the dts fixup file.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>
